firebase-query.html

```diff
@@ -280,18 +280,31 @@
 
         __queryChanged: function(query, oldQuery) {
           if (oldQuery) {
+            oldQuery.off('value', this.__onFirebaseValue, this);
             oldQuery.off('child_added', this.__onFirebaseChildAdded, this);
             oldQuery.off('child_removed', this.__onFirebaseChildRemoved, this);
             oldQuery.off('child_changed', this.__onFirebaseChildChanged, this);
             oldQuery.off('child_moved', this.__onFirebaseChildMoved, this);
 
             this.syncToMemory(function() {
+              this.__map = {};
               this.set('data', this.zeroValue);
             });
           }
 
           if (query) {
-            query.on('child_added', this.__onFirebaseChildAdded, this.__onError, this);
+            /* The 'value' event doesn't order the data, so we have to
+             * revert to the slow path if ordering is requested
```

Comment:
The DataSnapshot provided to the 'value' event is in fact ordered.  But .val() returns it as a plain JS Object, which is not ordered.  To preserve ordering, you can iterate the DataSnapshot using the .forEach() method.